### DISCUSSION TITLE
🧱 🚸 ✨ Plugin System

### DISCUSSION THREAD

**Discussion by tabaktoni** [seanjameshan/starknet.js#567] [Ideas]
**System Core:** Defines the minimum functionalities of the system.
**System Services:** Extended functionalities from the system implemented via plugins.
**Plugin Manager:** Manage the plugin’s lifecycle. It handles the registration and loading of the plugins.
**Plugin:** Represents an independent functionality executed in the context of a system service that extends the system functionality. All the system plugins follow a standard interface.

Define Lifecycle and allow plugins to cross communicate true events.
The goal is to contain core functionality inside starknet.js without bloating it with: good to have, easy to use, extras, cli, 3th party libs etc, but to allow developers to use the core system and extend its functionality for general purpose or specific contract implementations with predefined tools.

Plugin development should be blueprinted (registration, injection, or event base)
Plugin developers should be able to extend interfaces.

Examples:
Vue Plugins: https://vuejs.org/guide/reusability/plugins.html

User should install it like ex:
``npm i starknetjs @starknetjs/chainid @starknetjs/cli @starknetjs/easy-starknet @starknetjs/zeta-account-abstraction``

This is a general idea, I don't have a specific implementation in mind but it could be explored.
Please give your feedback on this.

**Comment by wraitii** [comment]
Some thoughts:

 - Extending functionality of Starknet.js without having to change/wrap the class interfaces (Account, Provider) could be nice. One example is addresses, most functions expect `string`, but that could be changed to accept different types-that-are-convertible-to-an-address, and then calling a `getAdressFrom` function inside the starknet.js boundary. Could be used with Starknet ID, or with Dapp specific objects, so you could call `transfer(userAccountA, userAccountB, token)` instead of having to do `transfer(userAccountA.getaddress(), ...)`.
 This could be implemented by having free-functions for these 'resolvers' that could be extended.

 - Adding new functionality to existing classes would be useful, such as a `multicall` function using e.g. Argent X's multi call. Doable in javascript by changing the prototype at runtime, but I'm not entirely sure how to support this neatly in Typescript.

 - There's a lot that can be done without extending the existing classes, by adding new utility functions and such. Those won't need any particular work.


The `Vue` plugin approach is rather straightforward, just an 'install' function and a call to a 'use' function. This is maybe underspecified but I would agree intuitively that there is little reason to limit things in javascript.

**Comment by dhruvkelawala** [comment]
I like the idea of having modular code in starknet.js and I don't see any problems in doing it except the maintenance cost would be high. I would suggest having `@starknet/core` as the main library. This would need a proper discussion but in general, it's a great idea

**Comment by tabaktoni** [comment]
Done in Beta will be part of the starknetjs v10
https://github.com/starknet-io/starknet.js/pull/1582
